### ISSUE TITLE
Tab completion for neovim commands still seem to be broken.

### ISSUE THREAD

**Issue by beune** [hrsh7th/nvim-cmp#874]
### FAQ

- [X] I have checked the [FAQ](https://github.com/hrsh7th/nvim-cmp#faq) and it didn't resolve my problem.

### Issues

- [X] I have checked [existing issues](https://github.com/hrsh7th/nvim-cmp/issues) and there are no open or closed issues with the same problem.

### Neovim Version

0.7.0

### Minimal reproducible config

```.lua
vim.opt.completeopt= "menu,menuone,noselect"

local has_words_before = function()
  local line, col = unpack(vim.api.nvim_win_get_cursor(0))
  return col ~= 0 and vim.api.nvim_buf_get_lines(0, line - 1, line, true)[1]:sub(col, col):match("%s") == nil
end

local M = {}
M.config = function()
  local cmp = require'cmp'
  local luasnip = require'luasnip'
  cmp.setup({
    snippet = {
      expand = function(args)
        require('luasnip').lsp_expand(args.body)
      end,
    },
    mapping = {
      ["<Tab>"] = cmp.mapping(function(fallback)
        if cmp.visible() then
          cmp.select_next_item()
        elseif luasnip.expand_or_jumpable() then
          luasnip.expand_or_jump()
        elseif has_words_before() then
          cmp.complete()
        else
          fallback()
        end
      end, { "i", "s" }),
      ["<S-Tab>"] = cmp.mapping(function(fallback)
        if cmp.visible() then
          cmp.select_prev_item()
        elseif luasnip.jumpable(-1) then
          luasnip.jump(-1)
        else
          fallback()
        end
      end, { "i", "s" }),
      ['<C-b>'] = cmp.mapping(cmp.mapping.scroll_docs(-4), { 'i', 'c' }),
      ['<C-f>'] = cmp.mapping(cmp.mapping.scroll_docs(4), { 'i', 'c' }),
      ['<C-Space>'] = cmp.mapping(cmp.mapping.complete(), { 'i', 'c' }),
      ['<C-y>'] = cmp.config.disable, -- Specify `cmp.config.disable` if you want to remove the default `<C-y>` mapping.
      ['<C-e>'] = cmp.mapping({
        i = cmp.mapping.abort(),
        c = cmp.mapping.close(),
      }),
      ['<CR>'] = cmp.mapping.confirm({ select = true }), -- Accept currently selected item. Set `select` to `false` to only confirm explicitly selected items.
    },
    sources = cmp.config.sources({
      { name = 'nvim_lsp' },
      { name = 'luasnip' },
      { name = 'path' },
      { name = 'buffer' },
    })
  })

  -- Set configuration for specific filetype.
  cmp.setup.filetype('gitcommit', {
    sources = cmp.config.sources({
      { name = 'cmp_git' }, -- You can specify the `cmp_git` source if you were installed it.
    }, {
      { name = 'buffer' },
    })
  })

  -- Use buffer source for `/` (if you enabled `native_menu`, this won't work anymore).
  cmp.setup.cmdline('/', {
    sources = {
      { name = 'buffer' }
    }
  })

  -- Use cmdline & path source for ':' (if you enabled `native_menu`, this won't work anymore).
  cmp.setup.cmdline(':', {
    sources = cmp.config.sources({
      { name = 'path' }
    }, {
      { name = 'cmdline' }
    })
  })
end

return M
```

### Description

Tab completion for neovim commands still don't seem to work. Related: #740.

### Steps to reproduce

Hit `:`, type something half-done, and then try `<Tab>` to auto-complete.

### Expected behavior

Can

### Actual behavior

Can't

### Additional context

None

**Comment by Shougo** [comment]
> Hit `:`, type something half-done, and then try `<Tab>` to auto-complete.

Well, it works for me.

Please see my minimal configuration.

```vim
set rtp+=~/src/nvim-cmp
set rtp+=~/src/cmp-buffer
set rtp+=~/src/cmp-nvim-lsp
set rtp+=~/src/nvim-lspconfig
set rtp+=~/src/cmp-cmdline
set rtp+=~/src/LuaSnip

lua <<EOF
vim.opt.completeopt= "menu,menuone,noselect"

local has_words_before = function()
  local line, col = unpack(vim.api.nvim_win_get_cursor(0))
  return col ~= 0 and vim.api.nvim_buf_get_lines(0, line - 1, line, true)[1]:sub(col, col):match("%s") == nil
end

local cmp = require'cmp'
local luasnip = require'luasnip'
cmp.setup({
  snippet = {
    expand = function(args)
      require('luasnip').lsp_expand(args.body)
    end,
  },
  mapping = {
    ["<Tab>"] = cmp.mapping(function(fallback)
      if cmp.visible() then
        cmp.select_next_item()
      elseif luasnip.expand_or_jumpable() then
        luasnip.expand_or_jump()
      elseif has_words_before() then
        cmp.complete()
      else
        fallback()
      end
    end, { "i", "s" }),
    ["<S-Tab>"] = cmp.mapping(function(fallback)
      if cmp.visible() then
        cmp.select_prev_item()
      elseif luasnip.jumpable(-1) then
        luasnip.jump(-1)
      else
        fallback()
      end
    end, { "i", "s" }),
    ['<C-b>'] = cmp.mapping(cmp.mapping.scroll_docs(-4), { 'i', 'c' }),
    ['<C-f>'] = cmp.mapping(cmp.mapping.scroll_docs(4), { 'i', 'c' }),
    ['<C-Space>'] = cmp.mapping(cmp.mapping.complete(), { 'i', 'c' }),
    ['<C-y>'] = cmp.config.disable,
    ['<C-e>'] = cmp.mapping({
      i = cmp.mapping.abort(),
      c = cmp.mapping.close(),
    }),
    ['<CR>'] = cmp.mapping.confirm({ select = true }),
  },
  sources = cmp.config.sources({
    { name = 'nvim_lsp' },
    { name = 'luasnip' },
    { name = 'path' },
    { name = 'buffer' },
  })
})

-- Set configuration for specific filetype.
cmp.setup.filetype('gitcommit', {
  sources = cmp.config.sources({
    { name = 'cmp_git' }, -- You can specify the `cmp_git` source if you were installed it.
  }, {
    { name = 'buffer' },
  })
})

-- Use buffer source for `/` (if you enabled `native_menu`, this won't work anymore).
cmp.setup.cmdline('/', {
  sources = {
    { name = 'buffer' }
  }
})

-- Use cmdline & path source for ':' (if you enabled `native_menu`, this won't work anymore).
cmp.setup.cmdline(':', {
  sources = cmp.config.sources({
    { name = 'path' }
  }, {
    { name = 'cmdline' }
  })
})
EOF

runtime! after/plugin/*.lua
```

**Comment by Shougo** [comment]
What is your desired behavior?  It is not easy to understand for me.

**Comment by hrsh7th** [comment]
> You must provide a working config based on [this](https://github.com/hrsh7th/nvim-cmp/blob/main/utils/vimrc.vim). Not part of config.

Please rrspect the issue template.

**Comment by beune** [comment]
> > Hit `:`, type something half-done, and then try `<Tab>` to auto-complete.
> 
> Well, it works for me.
> 
> Please see my minimal configuration.
> 
> ```viml
> set rtp+=~/src/nvim-cmp
> set rtp+=~/src/cmp-buffer
> set rtp+=~/src/cmp-nvim-lsp
> set rtp+=~/src/nvim-lspconfig
> set rtp+=~/src/cmp-cmdline
> set rtp+=~/src/LuaSnip
> 
> lua <<EOF
> vim.opt.completeopt= "menu,menuone,noselect"
> 
> local has_words_before = function()
>   local line, col = unpack(vim.api.nvim_win_get_cursor(0))
>   return col ~= 0 and vim.api.nvim_buf_get_lines(0, line - 1, line, true)[1]:sub(col, col):match("%s") == nil
> end
> 
> local cmp = require'cmp'
> local luasnip = require'luasnip'
> cmp.setup({
>   snippet = {
>     expand = function(args)
>       require('luasnip').lsp_expand(args.body)
>     end,
>   },
>   mapping = {
>     ["<Tab>"] = cmp.mapping(function(fallback)
>       if cmp.visible() then
>         cmp.select_next_item()
>       elseif luasnip.expand_or_jumpable() then
>         luasnip.expand_or_jump()
>       elseif has_words_before() then
>         cmp.complete()
>       else
>         fallback()
>       end
>     end, { "i", "s" }),
>     ["<S-Tab>"] = cmp.mapping(function(fallback)
>       if cmp.visible() then
>         cmp.select_prev_item()
>       elseif luasnip.jumpable(-1) then
>         luasnip.jump(-1)
>       else
>         fallback()
>       end
>     end, { "i", "s" }),
>     ['<C-b>'] = cmp.mapping(cmp.mapping.scroll_docs(-4), { 'i', 'c' }),
>     ['<C-f>'] = cmp.mapping(cmp.mapping.scroll_docs(4), { 'i', 'c' }),
>     ['<C-Space>'] = cmp.mapping(cmp.mapping.complete(), { 'i', 'c' }),
>     ['<C-y>'] = cmp.config.disable,
>     ['<C-e>'] = cmp.mapping({
>       i = cmp.mapping.abort(),
>       c = cmp.mapping.close(),
>     }),
>     ['<CR>'] = cmp.mapping.confirm({ select = true }),
>   },
>   sources = cmp.config.sources({
>     { name = 'nvim_lsp' },
>     { name = 'luasnip' },
>     { name = 'path' },
>     { name = 'buffer' },
>   })
> })
> 
> -- Set configuration for specific filetype.
> cmp.setup.filetype('gitcommit', {
>   sources = cmp.config.sources({
>     { name = 'cmp_git' }, -- You can specify the `cmp_git` source if you were installed it.
>   }, {
>     { name = 'buffer' },
>   })
> })
> 
> -- Use buffer source for `/` (if you enabled `native_menu`, this won't work anymore).
> cmp.setup.cmdline('/', {
>   sources = {
>     { name = 'buffer' }
>   }
> })
> 
> -- Use cmdline & path source for ':' (if you enabled `native_menu`, this won't work anymore).
> cmp.setup.cmdline(':', {
>   sources = cmp.config.sources({
>     { name = 'path' }
>   }, {
>     { name = 'cmdline' }
>   })
> })
> EOF
> 
> runtime! after/plugin/*.lua
> ```

I believe we have  the same mappings for Tab and S-Tab.

**Comment by beune** [comment]
> > You must provide a working config based on [this](https://github.com/hrsh7th/nvim-cmp/blob/main/utils/vimrc.vim). Not part of config.
> 
> Please respect the issue template.

My config is spread over multiple files:

```lua
return require('packer').startup(function()
  use {
      "hrsh7th/nvim-cmp",
      config = function()
          require("core.plugins.nvim-cmp").config()
      end
  }
  use 'saadparwaiz1/cmp_luasnip'
  use 'hrsh7th/cmp-nvim-lsp'
  use 'hrsh7th/cmp-path'
end
```
```lua
local servers = { 'pyright', 'vuels', 'bashls', 'eslint', 'ltex', 'sumneko_lua' }
for _, lsp in pairs(servers) do
  local capabilities = vim.lsp.protocol.make_client_capabilities()
  capabilities = require('cmp_nvim_lsp').update_capabilities(capabilities)
  require('lspconfig')[lsp].setup {
    on_attach = on_attach,
    flags = {
      -- This will be the default in neovim 0.7+
      debounce_text_changes = 150,
    },
    capabilities = capabilities,
  }
end
```
Please see previous comment for my `nvim-cmp` config.

**Comment by Shougo** [comment]
So, you need to minimal config.
We cannot reproduce your problem.

**Comment by beune** [comment]
> So, you need to create minimal config. We cannot reproduce your problem.

I used your minimal config, I still have the same problem. Not sure what other info I should provide.

**Comment by beune** [comment]
Fixed: turns out removing `{name = 'path'}` from
```lua
cmp.setup.cmdline(':', {
  sources = cmp.config.sources({
    { name = 'path' }
  }, {
    { name = 'cmdline' }
  })
})
```
Did the trick.

**Comment by feakuru** [comment]
FYI: 

> removing `{name = 'path'}`

worked for me with the default NVChad config. The commands were not showing on Tab before, now they are.

**Comment by kaiyulee** [comment]
```lua
-- Use buffer source for `/` and `?` (if you enabled `native_menu`, this won't work anymore).
cmp.setup.cmdline({ '/', '?' }, {
    -- mapping = cmp.mapping.preset.cmdline(),
    sources = {
        { name = 'buffer' }
    }
})

-- Use cmdline & path source for ':' (if you enabled `native_menu`, this won't work anymore).
cmp.setup.cmdline(':', {
    -- mapping = cmp.mapping.preset.cmdline(),
    sources = cmp.config.sources({
        { name = 'cmdline' }
    })
})


```

need to comment out the ` -- mapping = cmp.mapping.preset.cmdline(),`, then it works.

**Comment by phdah** [comment]
Quite a lot later, but after reading this and tinkering with it, I found this to work
```lua
cmp.setup.cmdline(':', {
    sources = cmp.config.sources({
        { name = 'cmdline' }, -- Priority 1
        { name = 'path' },    -- Priority 2
    })
})
```
By changing the ordering of these, it seems to _solve_ the problem with messing up the `tab` in command line since it first tries to complete any `command`, followed by trying to complete with a `path`.

If anyone has a link to another issue or whatever where this could be interesting to share, please let me know.

**Comment by Andrew15-5** [comment]
Today I said "enough is enough" after having my command mode tab completion killed for the nth time. Luckily, I remembered the key things I did in 1 or 2 minutes of opening the Neovim. Then I quickly found out that after searching for anything (`/`) the tab stops working (`^I`). The weird thing is that I didn't set up anything for the `:`, but I did have one setting for the `/`. Soon after I stumbled upon [this](https://neovim.discourse.group/t/tab-compleation-in-command-mode-stops-working/3094) and [this](https://github.com/neovim/neovim/issues/24611) which didn't have anything concrete. And now finally I'm here.

I had this set (which _sometimes_ prevents me from scrolling up the history log with arrow up, btw):

```lua
require("cmp").setup.cmdline("/", {
  mapping = cmp.mapping.preset.cmdline(),
  sources = cmp.config.sources({
    { name = "nvim_lsp_document_symbol" },
  }, {
    { name = "buffer" },
  }),
})
```

And after removing the `mapping` (as was said by @kaiyulee) this issue went away.

Now, after quickly scrolling through the readme file, I found this exact block of code:

```lua
cmp.setup.cmdline({ "/", "?" }, {
  mapping = cmp.mapping.preset.cmdline(),
  sources = {
    { name = "buffer" },
  },
})
```

Which in fact also breaks the command mode tab completion (because of the same `mapping`). And so I thought that here would be a good place to ask if @hrsh7th or anyone else knows why this happens and why the code that creates this issue/bug is in the readme? I think that either this issue should be reopened or the code snippet changed/removed from the readme. It does say that we all should use/manage our own keymaps, but this doesn't mean that the provided keymaps (`cmp.mapping.*`) should break things. Maybe there is some note about this or no one simply have time/desire to fix this? Just curious.

There is also this one

```help
Usage                                                                *cmp-usage*

A recommended configuration can be found below.
  NOTE:
    1. You must provide a `snippet.expand` function.
    2. `cmp.setup.cmdline` won't work if you use the `native` completion menu.
    3. You can disable the `default` options by specifying `cmp.config.disable` value.
 ...
window.{completion,documentation}.border~
  `string | string[] | nil`
  Border characters used for the completion popup menu when |experimental.native_menu| is disabled.
  See |nvim_open_win|.
```

And something about "native menu" (I'm not that deep into Neovim to understand what this is, even though I've been using it for several years). `:h nvim_open_win` don't have anything about "native".
